Small GUI Improvements

Pritchy96 opened the pull request:
A bunch of small GUI improvements: 

- Model Browser Preview can now be resized by clicking and dragging the column divider. Previously this divider could be clicked and dragged but would snap back on mouse release. 
- Texture Browser Menu (left column) resized to fit content, scales with UI scale. 
- Texture Browser and Model grid now have their Imagebuttons centered in each grid column.   
- The tools menu now no longer crops the tool buttons on certain GUI scales – this isn't perfect, there is some x offset that I can't figure out on these, but it's better. 
- Model class moved to show in the inspector instead of showing up in the grid. 

I'm sure most of this GUI will change, but this was just a few touch-ups to get some experience in ImGui.

Cheers,
Tom

chaoticgd commented:
> Model class moved to show in the inspector instead of showing up in the grid.

It's a bit ugly as it stands, but I put the model class in the grid so it could more easily be used as a reference sheet of sorts, so I'd like to keep that for now.

chaoticgd commented:
Also the model name isn't always the class. Look in level_impl.cpp and armor_archive.cpp. The code for that is a bit messy right now.

chaoticgd commented:
Other than that it seems work well, though it'd be good if you could enforce some kind of minimum column width.

chaoticgd commented:
I've just also noticed that the tools menu looks a bit borked now. It scales a bit better but looks broken at 1x scale.

![Screenshot_20210428_023305](https://user-images.githubusercontent.com/43898262/116332668-16a3d480-a7ca-11eb-84dd-b47a38f03a84.png)

Pritchy96 commented:
> > Model class moved to show in the inspector instead of showing up in the grid.
> 
> I put the model class in the grid so it could more easily be used as a reference sheet of sorts, so I'd like to keep that for now.

Yeah, totally valid. I assumed you might want to keep it, but though I'd suggest an alternative regardless. Do you think it's worth keeping it in the inspector, even if it's added back to the grid, just to ensure all the relevant information is kept in there? If so, would a better name just be "Class"? I will start doing some work familiarising myself with terminology and file structures as I continue to contribute.


> Other than that it seems work well, though it'd be good if you could enforce some kind of minimum column width.

This is a good idea, I'm not sure if ImGui supports column collapsing (i.e collapse to 0 width, but with a tab header to click to re-open the column at minimum width), but that could be good to add so a user can maximise the size of a column if they so wish.


> I've just also noticed that the tools menu looks a bit borked now. It scales a bit better but looks broken at 1x scale.
> 
> ![Screenshot_20210428_023305](https://user-images.githubusercontent.com/43898262/116332668-16a3d480-a7ca-11eb-84dd-b47a38f03a84.png)

Hmm, yeah this is what I meant by it not being perfect. I spent a decent while trying to figure out what the padding was on this and didn't quite manage to figure it out. My angle on this is that, prior to this change, higher scaling rates meant the tool window was actually being clipped, which looks even worse. It's worth noting that a 2x scale is more or less essential at 4k resolutions (and we should probably look at supporting higher scaling rates still) Happy to revert this change though, any UI stuff is a subjective mess anyway :)

I've got one additional GUI change (A preview image in the inspector, similar to what's in the model viewer grid, if a single moby type is selected) which I will append to this rather than put in a separate PR.  

Cheers! 
Tom

chaoticgd commented:
> Do you think it's worth keeping it in the inspector, even if it's added back to the grid, just to ensure all the relevant information is kept in there?

I'd prefer not to since in the case of non-level mobies it's incorrect.

> I will start doing some work familiarising myself with terminology and file structures as I continue to contribute.

The whole thing with the class being stored as the stream name is a bodge anyway, I'll probably rip it out at some point.

> This is a good idea, I'm not sure if ImGui supports column collapsing (i.e collapse to 0 width, but with a tab header to click to re-open the column at minimum width), but that could be good to add so a user can maximise the size of a column if they so wish.

Actually don't bother with that. I'm planning on replacing the model browser GUI at some point anyway.

> Hmm, yeah this is what I meant by it not being perfect. I spent a decent while trying to figure out what the padding was on this and didn't quite manage to figure it out. My angle on this is that, prior to this change, higher scaling rates meant the tool window was actually being clipped, which looks even worse. It's worth noting that a 2x scale is more or less essential at 4k resolutions (and we should probably look at supporting higher scaling rates still) Happy to revert this change though, any UI stuff is a subjective mess anyway :)

This expression seems to work much better:

> ImGuiStyle& style = ImGui::GetStyle();
> ImGui::SetNextWindowSize(ImVec2(32 + ((style.FramePadding.x + style.WindowPadding.x) * 2), view->Size.y));

From the implementation of ImageButton I gather it looks like this:
> | WindowPadding | Frame Padding | Icon | Frame Padding | WindowPadding |

Of course, then you have the problem where the dock area doesn't move with it, so you'll have to modify gui::begin_docking. Maybe compute the size in gui::render and pass it down.

chaoticgd commented:
Might be a good idea to scale the icon as well.

chaoticgd commented:
Also I just noticed you never added your name to the about screen. You can do that along with all of the other changes requested.

Pritchy96 commented:
> Also I just noticed you never added your name to the about screen. You can do that along with all of the other changes requested.

Oh neat - will do. I'll get round to these changes hopefully this week - sorry for the radio silence, I've been preoccupied moving house!

Pritchy96 commented:
Okay, I've made further changes per our previous discussions: 

- Reverted change to add model class to inspector. 
- Reverted change to remove model class to grid.
      - Part of the reason I removed this in the first place is because it was getting clipped on certain column widths:  

    
     ![image](https://user-images.githubusercontent.com/4840175/119274706-29fb6180-bc09-11eb-9c56-8054080c091a.png)


     - To try and alleviate this, I've put the index above the preview and the model class below: 

    
    ![image](https://user-images.githubusercontent.com/4840175/119274695-21a32680-bc09-11eb-9412-b5f6199bae2c.png)


    - If this isn’t to your liking, I'm happy to revert, I personally think it looks neater. 

- Added my name to the contributors section of the "About" dialogue. 

- Added a model preview to the inspector window in the 3D view tab. This is particularly useful when selecting from the 'mobies' tab.  

    ![image](https://user-images.githubusercontent.com/4840175/119274648-eb65a700-bc08-11eb-951e-8a8a53fc05d1.png)

     - If two separate types of moby, or a non moby are selected, nothing preview is rendered: 

    ![image](https://user-images.githubusercontent.com/4840175/119274655-fb7d8680-bc08-11eb-8272-8af3e18e7d83.png)

    - This is using code nabbed from the model preview - I did plan at one point to pull that code, as well as all other render grid/previews/etc code, out to a generic render helper class and remove the code reuse, but I thought that's probably something better suited to another PR after some discussion with yourself. 

- Minimum Column Width on model browser page.  

- I now scale the tools menu icons along with the tools menu width. I banged my head against a wall for longer than I'd care to admit, but this seems to be a good solution. Personally I think the Icons are a little large relative to any given scale value, but that's personal preference. 

Cheers,
Tom

chaoticgd commented:
The addition preview seems pretty good, even more so because the existing selection logic is quite dodgy.

> This is using code nabbed from the model preview - I did plan at one point to pull that code, as well as all other render grid/previews/etc code, out to a generic render helper class and remove the code reuse, but I thought that's probably something better suited to another PR after some discussion with yourself.

I don't see any reason to add a class for this, it seems like a good job for a function. Maybe put it in renderer.cpp, call it render_single_moby_to_texture and have it take a pointer to a texture handle (the output), the moby model and an instance of view_params as an argument, or something like that. We could leave it for a separate PR, although I don't see reason why that should be done.

> Personally I think the Icons are a little large relative to any given scale value, but that's personal preference.

A compact mode would be a good addition in the future. I originally made them that big as a reference to Valve's Hammer editor.

Looks mostly good now. Tell me you want to make these changes in this PR, or if you want me to merge this as-is.

chaoticgd commented:
Or even better than a render_single_moby_to_texture function, have a function that can be composed with the existing render_to_texture function instead of wrapping. So:

> render_to_texture(&tex, [&preview]() { draw_single_moby(view_params); })

chaoticgd commented:
Maybe remove the message you get when a non-moby is selected and instead add a line that just states what type of thing actually is selected at any given time e.g.

> 1 Moby

> 5 Mobies

> 2 Entities

Something short like that would be good I think.

chaoticgd commented:
Alright, merging now.
